framework/src/bcs/ADNodalBC.C

```diff
@@ -67,148 +68,92 @@ conversionHelper(const libMesh::VectorValue<ADReal> & value, const unsigned int
-                             cached_rows[i],
-                             conversionHelper(residual, i).derivatives()[cached_rows[i]],
-                             tag_id);
+  const auto residual = computeQpResidual();
+
+  processResidual(residual, dof_indices);
+  processJacobian(residual, dof_indices);
 }
 
 template <typename T>
 void
 ADNodalBCTempl<T>::computeOffDiagJacobian(const unsigned int jvar_num)
 {
+  // Only need to do this once because AD does all the derivatives at once
   if (jvar_num == _var.number())
     computeJacobian();
-  else
-  {
-    const std::vector<dof_id_type> & cached_rows = _var.dofIndices();
-    if (cached_rows.empty())
-      return;
-
-    auto residual = computeQpResidual();
-
-    mooseAssert(cached_rows.size() <= _set_components.size(),
-                "The number of dof indices must be less than the number of settable components");
-
-    // Note: this only works for Lagrange variables...
-    dof_id_type cached_col = _current_node->dof_number(_sys.number(), jvar_num, 0);
-
-    // Cache the user's computeQpJacobian() value for later use.
-    for (auto tag : _matrix_tags)
-      if (_sys.hasMatrix(tag))
-        for (std::size_t i = 0; i < cached_rows.size(); ++i)
-          if (_set_components[i])
-          {
-            _fe_problem.assembly(0, _sys.number())
-                .cacheJacobian(cached_rows[i],
-                               cached_col,
-                               conversionHelper(residual, i).derivatives()[cached_col],
-                               tag);
-          }
-  }
 }
 
 template <typename T>
 void
-ADNodalBCTempl<T>::computeOffDiagJacobianScalar(unsigned int jvar)
+ADNodalBCTempl<T>::computeOffDiagJacobianScalar(unsigned int)
 {
-  auto residual = computeQpResidual();
-  const std::vector<dof_id_type> & cached_rows = _var.dofIndices();
-
-  mooseAssert(cached_rows.size() <= _set_components.size(),
-              "The number of dof indices must be less than the number of settable components");
-
-  std::vector<dof_id_type> scalar_dof_indices;
-
-  _sys.dofMap().SCALAR_dof_indices(scalar_dof_indices, jvar);
-
-  // Our residuals rely on returning a single scalar and we don't provide any arguments to
-  // computeQpResidual so I think it only makes sense to assume that our SCALAR variable should be
-  // order one
-  mooseAssert(scalar_dof_indices.size() == 1,
-              "ADNodalBC only allows coupling of first order SCALAR variables");
-
-  // Cache the user's computeQpJacobian() value for later use.
-  for (auto tag : _matrix_tags)
-    if (_sys.hasMatrix(tag))
-      for (std::size_t i = 0; i < cached_rows.size(); ++i)
-        if (_set_components[i])
-        {
-          _fe_problem.assembly(0, _sys.number())
-              .cacheJacobian(cached_rows[i],
-                             scalar_dof_indices[0],
-                             conversionHelper(residual, i).derivatives()[scalar_dof_indices[0]],
-                             tag);
-        }
```

Comment:
why did we have all this code ?
were scalar variables not supported with AD at the time? you added it there?
https://github.com/idaholab/moose/pull/13704

Reply:
Yes I added that support in #13704 and up until now we've been specifically adding Jacobian elements corresponding to local field variable couplings and scalar variable coupling. With the change here to just add all coupling (local *and* nonlocal field variable coupling) then we no longer need to specifically add the scalar contributions